File: src/main/java/com/polygloat/security/api_key_auth/ApiAuthFilter.java
package com.polygloat.security.api_key_auth;

import com.polygloat.model.ApiKey;
import com.polygloat.service.ApiKeyService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Optional;

@Component
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class ApiAuthFilter extends OncePerRequestFilter {

    private final ApiKeyService apiKeyService;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        String keyParameter = request.getParameter("ak");

        if (keyParameter != null && !keyParameter.isEmpty()) {
            Optional<ApiKey> ak = apiKeyService.getApiKey(keyParameter);
            if (ak.isPresent()) {
                ApiKeyAuthenticationToken apiKeyAuthenticationToken = new ApiKeyAuthenticationToken(ak.get());
                SecurityContextHolder.getContext().setAuthentication(apiKeyAuthenticationToken);
            }
        }

        filterChain.doFilter(request, response);
    }

    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        String path = request.getRequestURI();
        return !path.startsWith("/uaa");
    }
}

File: src/main/kotlin/io/tolgee/security/api_key_auth/ApiAuthFilter.kt
package io.tolgee.security.api_key_auth

import io.tolgee.ExceptionHandlers
import io.tolgee.exceptions.PermissionException
import io.tolgee.service.ApiKeyService
import io.tolgee.service.SecurityService
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import org.springframework.web.method.HandlerMethod
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class ApiAuthFilter(
        private val apiKeyService: ApiKeyService,
        private val requestMappingHandlerMapping: RequestMappingHandlerMapping,
        private val securityService: SecurityService,
        private val exceptionHandlers: ExceptionHandlers
) : OncePerRequestFilter() {

    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
        if (this.isApiAccessAllowed(request)) {
            val keyParameter = request.getParameter("ak")
            if (keyParameter != null && !keyParameter.isEmpty()) {
                val ak = apiKeyService.getApiKey(keyParameter)
                if (ak.isPresent) {
                    val apiKeyAuthenticationToken = ApiKeyAuthenticationToken(ak.get())
                    SecurityContextHolder.getContext().authentication = apiKeyAuthenticationToken

                    val apiScopes = this.getAccessAllowedAnnotation(request)!!.scopes
                    try {
                        securityService.checkApiKeyScopes(setOf(*apiScopes), ak.get())
                    } catch (e: PermissionException) {
                        val errorResponseEntity = exceptionHandlers.handleServerError(e);
                        response.status = errorResponseEntity.statusCodeValue;
                        response.outputStream.print(errorResponseEntity?.body?.toString());
                        return;
                    }
                }
            }
        }
        filterChain.doFilter(request, response)
    }

    private fun isApiAccessAllowed(request: HttpServletRequest): Boolean {
        return this.getAccessAllowedAnnotation(request) != null
    }

    private fun getAccessAllowedAnnotation(request: HttpServletRequest): AllowAccessWithApiKey? {
        return (requestMappingHandlerMapping.getHandler(request)?.handler as HandlerMethod?)
                ?.getMethodAnnotation(AllowAccessWithApiKey::class.java) ?: return null
    }
}